doc/ezd.1
.TH EZD 1 local
.SH NAME
ezd \- easy drawing for programs on X11 displays
.SH SYNTAX
\fB ezd
.ft R
[ option ...] 
.SH DESCRIPTION
The 
\fBezd\fR
command reads commands on stdin and draws in one or more
windows on the X version 11 display defined by the environment variable
\fBDISPLAY\fR.  Event actions may be reported on stdout.
The command provides users with simple drawing, editing, and interaction
capabilities.  The capabilities of \fBezd\fR may be extended by adding
procedures written in Scheme.
.SH OPTIONS
.TP 10
\fB -i
invoke a Scheme interpreter.  In this environment
commands are sent to the \fBezd\fR drawing system
by calling the procedure \fIezd\fR.
.TP 10
\fB -l
.ft R
log incomming graphics commands to the file \fBezd.LOG\fR in the current
directory. Command logging may also be enabled by setting the environment
variable \fBEZDLOG\fR to the name for the log file.
.TP 10
\fB -p
.ft R
treat an end-of-file on stdin as a \fBpause\fR command.
.TP 10
\fB -nopixmap
.ft R
do not buffer window updates in a pixmap.  This may result in the screen
flickering when windows are updated.  This option may also be selected by
setting the environment variable \fBEZDNOPIXMAP\fR to any value.
.TP 10
\fB -s
.ft R
unless this flag, or the \fB-p\fR or \fB-i\fR flags are specified, the
\fBezd\fR process will ignore the keyboard interrupt signal.
.TP 10
\fB -sc\fIxx
.ft R
configure the heap.  See the man page for \fBsci\fR for details.
.SH GETTTING STARTED
All commands are in the form of Scheme lists. A semicolon ";" indicates the
start of a comment.  The comment continues until the end of the line.  Space,
tab, and newline characters may be used to separate
commands or items within commands. Upper and lower case
letters are treated identically except inside strings.  A '"' is
.eo
represented inside a string by '\"' and a '\' is represented by '\\'.
.ec \
Numeric values in commands may be entered as either integer or floating point
values.

Initially, try a few drawing commands.  If you'd like to print the results,
use the \fBpostscript\fR command.  For
multiple windows, or a different coordinate system, use the window
commands.  Finally, for more sophisticated effects, look at the editing and
interactive commands.

The drawing model supported by \fBezd\fR can be described as follows. Graphical
objects are drawn into drawings and drawings are displayed by mapping them onto
windows.  When drawing objects, their positions and
line widths are specified in the application's coordinate system.  These
coordinates and widths may be expressed in real numbers.  Objects are drawn 
using the \fIpainters algorithm\fR: objects are opaque and later drawn objects
can
obscure earlier objects. Drawings are only
visible when they are mapped into a window.  This mapping describes the
transformation of the application's coordinate system to the X coordinate
system.  Operations such as panning and zooming can be simply done by changing
the mapping of a drawing onto a window.  Multiple drawings may be displayed in
a window and multiple windows may display a drawing.

If the application doesn't initially explicitly create any windows or
drawings, then
the \fIcurrent drawing\fR is named \fIezd\fR and it is displayed in
a window named \fIezd\fR.  There is
a 1-1 mapping between the application's coordinate system and X's coordinate
system.  That is, positions and line widths are specified in pixels with the
x-axis going from left to right and the y-axis going from top to bottom on the
window.  The origin is in the upper left corner of the window.

In the command descriptions that follow, literal values are in bold face type
and arguments are in italics.  Optional items are surrounded by [], and ...
denotes sequences of zero or more items.
.SH DRAWING COMMANDS
Graphical objects are drawn in the \fIcurrent drawing\fR by the following
commands.
The \fIxy\fR coordinates and
object \fIwidth\fR and \fIheight\fR are specified in terms of the application's
coordinate system.  Lines are either minimal width, or \fIline-width\fR
units wide (see the \fBscale\fR command to understand how application
coordinates and line widths are translated to pixels).  Lines are solid unless
\fBdash\fR is specified.
Objects are drawn in the display's default black unless a
\fIcolor\fR is supplied.  Colors are specified by their X color name.  One
additional color, \fBclear\fR, is defined that allows the creation of
transparent objects.  This is typically used to make invisible objects that
are sensitive to events.
A few stipple patterns have been defined:  \fBs0\fR, \fBs1\fR, \fBs2\fR, ... 
\fBs16\fR, where the number is the number of pixels that are drawn in a
4-by-4 pattern.

\fB (point \fIx y [color]\fB)\fR

draws a point.

\fB (line \fIx1 y1 x2 y2 [line-width] [color] [\fBdash\fI]\fB)\fR

draws a line from \fIx1 y1\fR to \fIx2 y2\fR.

\fB (arc \fIx y width height angle1 angle2 [line-width] [color] [\fBdash\fI]\fB)\fR
.br
\fB (fill-arc \fIx y width height angle1 angle2 [color] [stipple]\fB)\fR
.br
\fB (pie-arc \fIx y width height angle1 angle2 [color] [stipple]\fB)\fR

draws an unfilled or filled arc, where \fIx\fR and \fIy\fR are the minimum
coordinates of the rectangle defining the arc. 
The arc's rectangle size is \fIwidth\fR by \fIheight\fR.  The arc starts at
\fIangle1\fR degrees and spans \fIangle2\fR degrees.  Positive angles are
measured in the same direction as moving from the positive x-axis to the
positive y-axis.  An arc drawn by \fBfill-arc\fR is closed by drawing a line
between the end points of the arc.  An arc drawn by \fBpie-arc\fR is closed by
drawing a line from each arc end point to the center of the rectangle defining
the arc.

\fB (rectangle \fIx y width height [line-width] [color] [\fBdash\fI]\fB)\fR
.br
\fB (fill-rectangle \fIx y width height [color] [stipple]\fB)\fR

draws an unfilled or filled rectangle, where \fIx\fR and \fIy\fR are the
minimum coordinates.

\fB (fill-polygon \fIx1 y1 x2 y2 ... xn yn [color] [stipple]\fB)\fR

draws a polygon with these points and fills it.

\fB (text \fIx y [width height center-x center-y] \fB"\fItext\fB"\fI [color] [stipple] [\fB"\fIfont\fB"\fI]\fB) \fR

writes text where \fIx\fR and \fIy\fR are the minimum coordinates of the
rectangle containing the text.  If width, height and positioning information
is supplied then that information describes the size of the rectangle and
how the text is to be placed in it.  If the text will not fit in that
rectangle when the drawing is scaled for display in a window, then no text
is displayed.  \fICenter-x\fR is specified as \fBleft\fR, \fBcenter\fR, or
\fBright\fR, and \fIcenter-y\fR is specified as \fBup\fR, \fBcenter\fR, or
\fBdown\fR.  Note too that the optional font name is specified by a string that
is the X font name.

\fB (quilt \fIx y width height columns rows color ... \fB"\fIsquares\fB")\fR
.br
\fB (quilt \fIx y width height columns rows color ... \fB#(\fIsquares\fB))\fR

draws a rectangle filled with a number of different colored rectangles.  The
minimum coordinates are \fIx\fR and \fIy\fR and the rectangle is \fIwidth\fR
wide and \fIheight\fR high.  The rectangle is in turn broken horizontally into
a number of \fIcolumns\fR and vertically into a number of \fIrows\fR.  The
square colors are defined in \fIsquares\fR which contains indexes into the list
of \fIcolors\fR.  When \fIsquares\fR is a string, each square's color is
defined by a letter, where A is the first color, B the second, etc.  A quilt
may have holes in it.  Each missing square is denoted by using a space as the
color index.  When more than 26 colors are desired, the squares are defined
by entries in a Scheme vector, where 0 is the first color, 1 the second, etc.
A hole is specified by #f.

\fB (bitmap \fIx y [width height] filename [color ...]\fB)\fR

draws a quilt defined in a file.  The file must contain an X11 bitmap,
a monochrome portable bitmap (PBM, type P1), a gray scale portable bitmap
(PGM, type P2), or a color portable bitmap (PPM, type P3).  If the width and
height of the quilt are not defined, they default to the dimensions of the
bitmap. X11 and monochrome
bitmaps are handled as follows.  If no color is suppled, then the bits that
are on are drawn in black, and the bits that are off are holes.  If one color
is supplied, then the bits that are on are drawn in that color, and the bits
that are off are holes.  When two colors are supplied, the bits that are on are
drawn in the first color, and the bits that are off are drawn in the second
color.  Gray scale bitmaps are drawn using a 101 step gray scale ramp when no
colors are supplied.  When colors are supplied, they define the gray scale ramp
that is to be used (darkest colors first).  Color bitmaps are drawn using the
RGB values contained in the map.
.SH COLORS
\fB(define-color \fInew-color \fB#x\fIRRGGBB\fB)\fR
.br
\fB(define-color \fInew-color hue saturation value\fB)\fR
.br
\fB(define-color \fInew-color color\fB)\fR

defines a new color, where the value of \fInew-color\fR is specified by the
RGB value encoded in a hexadecimal number \fB#x\fIRRGGBB\fR, by the HSV
value, or by the value of \fIcolor\fR.

\fB(define-stipple \fIstipple row-bit-value ...\fB)\fR

defines a new stipple.  Each \fIrow-bit-value\fR is an integer that defines
the bits in one row of the stipple from left-to-right.  The rows are specified
from top-to-bottom and 4, 8, or 16 rows must be specified.

Getting just the right color in drawings is a difficult process.  Variable
colors allow the user to define colors that can be adjusted.

\fB(define-variable-color \fIvariable-color \fB#x\fIRRGGBB\fB)\fR
.br
\fB(define-variable-color \fIvariable-color hue saturation value\fB)\fR
.br
\fB(define-variable-color \fIvariable-color color\fB)\fR

defines a variable color, where the initial value of \fIvariable-color\fR
is specified by the RGB value encoded in the hexadecimal number
\fB#x\fIRRGGBB\fR, the HSV value, or the value of \fIcolor\fR.

\fB(set-variable-color \fIvariable-color \fB#x\fIRRGGBB\fB)\fR
.br
\fB(set-variable-color \fIvariable-color hue saturation value\fB)\fR
.br
\fB(set-variable-color \fIvariable-color \fIcolor\fB)\fR

sets the value of \fIvariable-color\fR to the RGB value encoded in the
hexadecimal number \fB#x\fIRRGGBB\fR, the HSV value, or the value of
\fIcolor\fR.

\fB(edit-variable-color \fIvariable-color\fB)\fR

displays a color control panel with four sliders and a text input
area that allows the user to
modify \fIvariable-color\fR.  Rgb values are changed by moving the appropriate
slider.  Gray values are set by moving the gray slider.  When a slider moves,
the name of the color (found in /usr/local/lib/X11/rgb.txt) closest to the
value of the sliders and the offset from the slider values to it is displayed.
The color can also be set by typing a color name or hexadecimal value into the
input area.  For details on how to run sliders or use text input areas, see
the corresponding sections of this document.

\fB(get-color-value \fIcolor\fB)\fR

writes a record on stdout containing the RGB and HSV definitions of a color.
See the HANDLING EVENTS section for more details.
.SH DRAWINGS
Graphical objects are drawn into drawings that are displayed by mapping
them into windows.  If no drawings are defined, then the default drawing
\fIezd\fR is used.

\fB (set-drawing \fIdrawing\fB)\fR

sets the name of the \fIcurrent drawing\fR to \fIdrawing\fR.  All graphics
commands that follow pertain to that drawing.  Legal drawing names are any
Scheme symbol.

\fB (save-drawing)\fR

pushes the name of the \fIcurrent drawing\fR onto a stack of drawing names.

\fB (restore-drawing)\fR

pops the name of a drawing off the stack of drawing names and makes it the
\fIcurrent drawing\fR.
.SH EDITING DRAWINGS
Objects in drawings may be named, replaced, deleted or rearranged by the
following commands.

\fB (object \fIobject [drawing-command ...]\fB)\fR

draws a named object in the \fIcurrent drawing\fR.  Object names may be any
Scheme symbol except *.
If the object already exists,
then it is replaced by the new object.  If no drawing commands are specified,
the the object still exists, but with a null graphical representation.
The first time that a named object is drawn
assigns its position in the drawing hierarchy.  Whenever it is redrawn, it will
occupy the same position unless moved by a \fBfloat\fR or \fBsink\fR command,
or the entire drawing is cleared by a \fBclear\fR command.

\fB (float \fIobject\fB)\fR

move the named object to the top of the \fIcurrent drawing\fR.

\fB (float \fIobject1 object2\fB)\fR

move \fIobject1\fR on top of \fIobject2\fR in the \fIcurrent drawing\fR.

\fB (sink \fIobject\fB)\fR

move the named object to the bottom of the \fIcurrent drawing\fR.

\fB (sink \fIobject1 object2\fB)\fR

move \fIobject1\fR under \fIobject2\fR in the \fIcurrent drawing\fR.

\fB (draw-now)\fR

causes any buffered changes in drawings to be immediately reflected in windows
displaying
them.  Normally, \fBezd\fR will buffer changes until an appropriate time to
redraw the windows.  At that time, it will do a minimum of redrawing to make
the changes.

\fB (clear)
.ft R

clears the \fIcurrent drawing\fR and deletes the ordering and event actions
associated with named objects.
.SH WINDOW COMMANDS
Drawings are displayed when they are mapped into a window.  Windows are
created and drawings are mapped into windows by the commands found in this
section.

\fB (window \fIname [right down] width height [\fBfixed-size\fI] [\fBpoints\fI] [\fB"\fItitle\fB"\fI] [foreground [background]]\fB)\fR

defines a display window with the name \fIname\fR.
The location of the upper left corner of the window is \fIright\fR units right
and \fIdown\fR units down from the upper left corner of the display.
The window is \fIwidth\fR units wide and
\fIheight\fR units high.  Window location, width, and height are assumed to be
in pixels
unless the keyword \fBpoints\fR is supplied in which case they are in points
(~1/72 inches). The window may be resized unless \fBfixed-size\fR is specified.
The window and icon are titled \fB"\fItitle\fB"\fR.
The foreground color is either the display's default black or \fIforeground\fR.
The background color is either the display's default white or \fIbackground\fR.
Windows are only visible when there are one or more drawings displayed in
them.  If the window already exists when this command is issued, the existing
window with this name is deleted.

\fB (delete-window \fIwindow\fB)\fR

deletes the window named \fIwindow\fR.

\fB (save-cursor \fIwindow\fB) \fR

saves the current cursor on a cursor stack associated with the window
\fIwindow\fR.

\fB (set-cursor \fIwindow cursor\fB) \fR

sets the cursor for window \fIwindow\fR to \fIcursor\fR.  Cursor
names are as defined in <X11/cursorfont.h>.  The default cursor is XC_left_ptr.

\fB (restore-cursor \fIwindow\fB) \fR

pops a cursor from the cursor stack associated with
\fIwindow\fR and makes it the window's cursor.

\fB (overlay \fIwindow drawing [right down width height [\fBpoints\fI]]\fB) \fR

overlays the \fIdrawing\fR over the existing drawings in the \fIwindow\fR.
If the drawing is currently visible in the window,
it is moved to the new position.  There is
a 1-1 mapping between the application's coordinate system and X's coordinate
system.  That is, positions and line widths are specified in pixels with the
x-axis going from left to right and the y-axis going from top to bottom on the
window.  The origin is in the upper left corner of the window.
An optional clipping region may be specified to restrict the drawing to
a portion of the window. The location of the upper left corner of the clipping
region is \fIright\fR units right
and \fIdown\fR units down from the upper left corner of the window.
The clip region is \fIwidth\fR units wide and
\fIheight\fR units high. Clip region location, width, and height are assumed to
be in pixels
unless the keyword \fBpoints\fR is supplied in which case they are in points
(~1/72 inches).

\fB (underlay \fIwindow drawing [right down width height [\fBpoints\fI]]\fB) \fR

places the \fIdrawing\fR under the existing drawings in the \fIwindow\fR.
An optional clipping region may be specified to restrict the drawing to
a portion of the window. If the drawing is currently visible in the window,
it is moved to the new position.  There is
a 1-1 mapping between the application's coordinate system and X's coordinate
system.  That is, positions and line widths are specified in pixels with the
x-axis going from left to right and the y-axis going from top to bottom on the
window.  The origin is in the upper left corner of the window.
An optional clipping region may be specified to restrict the drawing to
a portion of the window. The location of the upper left corner of the clipping
region is \fIright\fR units right
and \fIdown\fR units down from the upper left corner of the window.
The clip region is \fIwidth\fR units wide and
\fIheight\fR units high. Clip region location, width, and height are assumed to
be in pixels
unless the keyword \fBpoints\fR is supplied in which case they are in points
(~1/72 inches).

\fB (delete-view \fIwindow drawing\fB) \fR

deletes the \fIdrawing\fR from the \fIwindow\fR.  If this is
the only drawing displayed in the window, then the window will no longer be
visible.

When a drawing is mapped onto a window, the default coordinate system is that
one drawing unit is one pixel.  This can be changed by the \fBorigin\fR
and \fBscale\fR commands.  The \fBorigin\fR command specifies the X
coordinates (in either pixels or points) of the origin of the application's
coordinate system.  The \fBscale\fR command provides the scale factors (in either pixels or points) for scaling the x and y coordinates and line widths. 
The coordinate transformations are:

.in +10
.br
X_x = drawing_x * scale_x + origin_x
.br
X_y = drawing_y * scale_y + origin_y
.br
X_line_width = drawing_line_width * scale_linewidth
.in -10

\fB(origin \fIwindow drawing x y [\fBpoints\fI]\fB) \fR

defines the X coordinates of the origin in either pixels (the
default) or points for \fIdrawing\fR mapped onto \fIwindow\fR.
In X the origin is in the upper left corner of the
window, the positive x-axis goes from left to right on the window, and the
positive y-axis goes from top to bottom of the window.

\fB(scale \fIwindow drawing scale-x scale-y scale-width [\fBpoints\fI]\fB)\fR

defines the scale factors for the x,y coordinates and line widths in
either pixels (the default) or points for \fIdrawing\fR mapped onto 
\fIwindow\fR.  Scale factors may be real numbers.

Using these two commands, one can map a drawing with a classical cartesian
coordinate system onto a 400 by 400 pixel window by the commands:
.nf

	(origin any-window a-drawing 200 200)
	(scale any-window a-drawing 1 -1 1)
.fi
.SH INTERACTIVE COMMANDS
Objects in drawings are made sensitive to events such as keyboard and mouse
actions by the commands in the sections that follow.  The initial commands
provide
a mechanism for monitoring simple events such as mouse clicks, mouse motion,
and keystrokes.  The application writer may choose to simply log the event to
the application, or they may supply a sophisticated event handler written in
Scheme.

In order to simplify application writing, \fBezd\fR provides commands for
creating a modest number of 
interface objects such as buttons, sliders, menus and text
drawings.  Each interface object consists of a graphical representation and
appropriate event handlers.  Such objects often need to be communicated with
once created.  For example, a slider object may wish to allow the user to get
or set the value of the slider.  To allow this, \fBezd\fR provides the
commands:  \fBget-attributes\fR to get an attribute value from and object,
and \fBset-attribute\fR to set the value of an object's attribute.

Finally, the user is encouraged to extend \fBezd\fR by writing their own
interface objects.  In many ways, this is a similar philosophy to Xt where one
is encouraged to reuse or extend existing widgets.  However, this is much
easier in \fBezd\fR as it provides a simpler and more powerful graphics and
event model than is provided by Xlib and Xt.  Enough tub beating, on to
events.
.SH HANDLING EVENTS
One or more \fBwhen\fR commands is used to make an object sensative to events.

\fB (when \fIobject event action\fB) \fR
.br
\fB (when \fIobject [keyboard-modifier ...] \fBbutton\fIb\fBdown\fI action\fB) \fR
.br
\fB (when \fIobject\fB * #f) \fR

defines an event handler for an object in the \fIcurrent drawing\fR.
\fIObject\fR is either the name of a specific object, or a \fB*\fR to
denote all objects in the drawing.  The type of event that the object is
sensitive to is defined by \fIevent\fR.  The event types are:
.nf

	\fBenter\fR		mouse entered the object
	\fBexit\fR		mouse exited the object
	\fBmotion\fR		mouse moved within the object
	\fBbutton\fIb\fBup\fR	mouse button number \fIb\fR came up. Buttons are numbered 1-5.
	\fBbutton\fIb\fBdown\fR	mouse button number \fIb\fR went down. This event may be
			preceeded by any combination of the keywords \fBshift\fR, \fBmeta\fR,
			\fBctrl\fR, and \fBlock\fR indicating that those keys must be down
			for the event to occur.
	\fBkeypress\fR	keyboard key pressed
	\fBkeyrelease\fR	keyboard key released
	\fBresize\fR		a window containing the drawing was resized.
	\fBexpose\fR		a window containing the drawing was exposed.
	\fBoverlay\fR		the drawing was overlayed or underlayed on a window.
	\fBvisible\fR		a change was made to the portion of the drawing that is visible.
	\fBget-attributes\fR	a message was sent to the object via the \fBget-attributes\fR command.
	\fBset-attributes\fR	a message was sent to the object via the \fBset-attributes\fR command.

.fi
The \fIaction\fR is either a Scheme expression to evaluate or a Scheme
procedure (that takes no arguments) to be called.
Once defined, event handlers remain until they are deleted by supplying
\fB#f\fR as the action, or the drawing is cleared by the \fBclear\fR command.
All event actions associated with an object may be deleted by supplying \fB*\fR
as the event type and \fB#f\fR as the action.

Programs using \fBezd\fR as a graphics server can log events to the application
by supplying \fB(log-event \fI[user-info ...]\fB)\fR as the \fIaction\fR.
Each event causes an
event record of the following form to be printed on stdout:

   \fB(\fIevent window drawing object drawing-x drawing-y mouse-x mouse-y [misc-info ...] [user-info ...]\fB)\\n \fR

where \fIevent\fR is the name of the event, \fIwindow\fR the name of the
window, and \fIobject\fR is the name of the object (or \fB*\fR if no object was
involved).  The current mouse position in the drawing is returned by
\fIdrawing-x\fR and \fIdrawing-y\fR and the current mouse position in the
window in the X coordinate system is returned
by \fImouse-x\fR and \fImouse-y\fR (these values may be invalid for resize,
expose, visible, and overlay events). The last items are optional event
specific
information returned in \fImisc-info\fR and caller supplied items in
\fIuser-info\fR.  The event
specific information is defined as follows:
.TP 10
\fB exit
.ft R
a list of the names of the window, drawing, and object now containing the
mouse cursor.  A window value equal to \fB#f\fR indicates that the mouse is in
a window not controlled by \fBezd\fR.
.TP 10
\fB button\fIb\fBdown
.ft R
a list of the keyboard modifier keys.
.TP 10
\fB keypress\fR and \fBkeyrelease
.ft R
a list of a string holding the ISO Latin-1 value of the key and a number
that is the X KeySym for the key.
.TP 10
\fB resize
.ft R
a list of four window measurements in pixels: old window width, old window height, new window width, and new window height.
.TP 10
\fB overlay
.ft R
a list defining the bounding box for the overlay measured in pixels in the X
coordinate system: x, y, width, and height.  If the drawing is no longer
visible in the window associated with the event, then \fB#f\fR is the value of
each coordinate.
.TP 10
\fB visible
.ft R
a list defining the bounding box for the portion of the drawing that is
visible in the window associated with the event in the drawing's coordinate
system: x, y, width, and height.  If the drawing is no longer visible in the
window associated with the event, then \fB#f\fR is the value of each
coordinate.
.TP 10
\fB get-attributes
a list of attributes.  Each attribute is either a symbol or a list
of a symbol and its optional arguments.  The event handler for this event must
return a list of values of the attributes.
.TP 10
\fB set-attributes
a list of the attributes and their new values.  Each attribute is a list
consisting of the attribute name followed by any arguments.
.TP 10
\fB color-value
a list of RGB and HSV values for a color.  The window, drawing,
object, and mouse coordinate values are invalid in this record.
.PP
The \fBresize\fR, \fBexpose\fR, \fBoverlay\fR, and \fBvisible\fR events
happen to all objects in the drawing.  When one of these events happen, the
event handler for all objects (\fIobject\fR = *) is called and then any object
specific event handlers are called.

Often an application wishes to detect that the user "clicked" a button on the
object.  While one can implement this by installing event handlers for object
entry, exit, button down, and button up; this functionality can be installed
with one command.

\fB (click \fIobject [keyboard-modifier ...] button-number action\fB) \fR

defines an action to be done when a button is clicked, i.e. pressed down,
possibly moved, and then released within the object in the \fIcurrent

ezd.scm
;;; ezd - easy drawing for X11.
;;;
;;; Batch command loop and Scheme command interpreter.

;*           Copyright 1990-1993 Digital Equipment Corporation
;*                         All Rights Reserved
;*
;* Permission to use, copy, and modify this software and its documentation is
;* hereby granted only under the following terms and conditions.  Both the
;* above copyright notice and this permission notice must appear in all copies
;* of the software, derivative works or modified versions, and any portions
;* thereof, and both notices must appear in supporting documentation.
;*
;* Users of this software agree to the terms and conditions set forth herein,
;* and hereby grant back to Digital a non-exclusive, unrestricted, royalty-free
;* right and license under any changes, enhancements or extensions made to the
;* core functions of the software, including but not limited to those affording
;* compatibility with other hardware or software environments, but excluding
;* applications which incorporate this software.  Users further agree to use
;* their best efforts to return to Digital any such changes, enhancements or
;* extensions that they make and inform Digital of noteworthy uses of this
;* software.  Correspondence should be provided to Digital at:
;* 
;*                       Director of Licensing
;*                       Western Research Laboratory
;*                       Digital Equipment Corporation
;*                       250 University Avenue
;*                       Palo Alto, California  94301  
;* 
;* This software may be distributed (but not offered for sale or transferred
;* for compensation) to third parties, provided such third parties agree to
;* abide by the terms and conditions of this notice.  
;* 
;* THE SOFTWARE IS PROVIDED "AS IS" AND DIGITAL EQUIPMENT CORP. DISCLAIMS ALL
;* WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF
;* MERCHANTABILITY AND FITNESS.   IN NO EVENT SHALL DIGITAL EQUIPMENT
;* CORPORATION BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
;* DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
;* PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
;* ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
;* SOFTWARE.

#;    (with xlib
	  struct
	  match
          commands
	  pps
	  ginfo
	  display
	  window
	  drawing
	  view
	  graphic
	  rectangle
	  line
	  text
	  arc
	  quilt
	  psdraw
	  events
	  interact
	  stringinput
	  popupmenu
	  buttons
	  slider
	  textree
	  jtextree
	  textdrawing
	  mouseedit
	  editcolor
	  transpbuttons)

(module ezd ()

(import scheme chicken data-structures)

(use scheme2c-compatibility xlib posix srfi-18 srfi-1 srfi-4 lolevel extras)

;; disable buffering so that select works correctly.
(set-buffering-mode! (current-input-port) #:none)
(set-buffering-mode! (current-output-port) #:none)

(include "scheme2c.scm")

;; Exports
(include "commands.sch")
(include "display.sch")
(include "window.sch")
(include "view.sch")
(include "drawing.sch")
(include "graphic.sch")
(include "psdraw.sch")
(include "events.sch")
(include "ezd.sch")
(include "interact.sch")

;; Actual code
(include "struct.scm")
(include "match.scm")
(include "commands.scm")
(include "ginfo.scm")
(include "display.scm")
(include "window.scm")
(include "drawing.scm")
(include "view.scm")
(include "graphic.scm")
(include "rectangle.scm")
(include "line.scm")
(include "text.scm")
(include "arc.scm")
(include "quilt.scm")
(include "psdraw.scm")
(include "events.scm")
(include "interact.scm")
(include "stringinput.scm")
(include "popupmenu.scm")
(include "buttons.scm")
(include "slider.scm")
(include "textree.scm")
(include "jtextree.scm")
(include "textdrawing.scm")
(include "mouseedit.scm")
(include "transpbuttons.scm")

;;; Version tag

(define *ezd-version* "15mar93jfb")

;;; The program accepts five command line arguments arguments:
;;;
;;;	-i		interactive Scheme interpreter
;;;	-l		log commands to ezd.LOG
;;;	-nopixmap	don't use a pixmap when updating windows
;;;	-p		treat an end-of-file on stdin as a pause command
;;;	-s		allow interrupt signal
;;;
;;; If the environment variable EZDLOG is set, then commands are logged to a
;;; file with EZDLOG's value as its name.  If the environment flag EZDNOPIXMAP
;;; is set, then pixmaps are not used for window updates.

(define ezd-done #f)

(define ezd-log #f)

(define env-ezdlog (or (getenv "EZDLOG") ""))
(define env-ezdnopixmap (or (getenv "EZDNOPIXMAP") ""))

(define nopixmap (if (eq? env-ezdnopixmap "") #f #t))

(define (read-eval-draw clargs)
    (define pause (member "-p" clargs))
    
    
    (set! ezd-log (if (or (member "-l" clargs) (not (equal? env-ezdlog "")))
		      (let ((port (open-output-file (if (equal? env-ezdlog "")
							"ezd.LOG"
							env-ezdlog))))
			   (format port ";;; *EZD-VERSION* ~a~%" *ezd-version*)
			   port)
		      #f))
    (set! nopixmap (or (member "-nopixmap" clargs) nopixmap))
    (if (member "-i" clargs)
	(ezd-repl)
	(let ((old-reset reset))
	     (if (and (not (member "-p" clargs)) (not (member "-s" clargs)))
		 (set-signal-handler! 2 #f))
	     (set! ezd-done #f)
	     (set! in-read-eval-draw #t)
	     (call-with-current-continuation
		 (lambda (return)
			 (set! reset (lambda () (return #t)))))
	     (let loop ((command (or ezd-done (ezd-read))))
		  (if ezd-log
		      (if (or ezd-done (eof-object? command))
			  (close-output-port ezd-log)
			  (begin (write command ezd-log)
				 (newline ezd-log))))
		  (if (or ezd-done (eof-object? command))
		      (begin (if (and (eof-object? command) pause)
				 (ezd '(pause)))
			     (ezd-reset))
		      (begin (ezd command)
			     (loop (or ezd-done (ezd-read))))))
	     (set! reset old-reset)
	     (set! ezd-done #f)
	     (set! in-read-eval-draw #f))))

;;; One or more ezd commands are executed by the following function.  Any
;;; graphical objects generated are added to the current drawing.  This
;;; procedure is the command interface for ezd from Scheme user programs.

(define (ezd . commands)
    (for-each
	(lambda (c)
		(let ((value (ezd-one c)))
		     (if (isa-graphic? value)
			 (drawing-add *current-drawing* value))))
	commands)
    #f)

;;; Execute a single ezd command and return its value.  This is not an external
;;; interface.

(define (ezd-one command)
    (if (pair? command)
	(let* ((x (assoc (car command) ezd-commands))
	       (template (if x (cadr x)))
	       (description (if x (caddr x)))
	       (action (if x (cadddr x))))
	      (when (and (not *display*)
			 (not (memq (car command) '(include quit scheme))))
		    (set! *display* (make-display ""))
		    (ezd '(set-drawing ezd))
		    (if (not (memq (car command)
				   '(window save-drawing set-drawing)))
			(ezd '(window ezd 400 400 points "ezd")
			     '(overlay ezd ezd))))
	      (if x
		  (let ((args (arg-parse template command)))
		       (if (eq? args #f)
			   (ezd-error 'ezd
			       "Illegal command: ~s~%          expected - ~a"
			       command description)
			   (apply action args)))
		  (ezd-error 'ezd "Unrecognized command: ~s" command)))
	(ezd-error 'ezd "Command is not a list: ~s" command)))

;;; The following procedure resets the entire drawing system.  All modules
;;; needing initialization in order to be rerun must define a reset procedure
;;; and have it called from here.

(define (ezd-reset)
    (ezd-module-init)
    (commands-module-init)
    (display-module-init)
    (window-module-init)
    (view-module-init)
    (drawing-module-init)
    (graphic-module-init)
    (events-module-init))

;;; The SCHEME command allows Scheme expressions to be evaluated within ezd.

(define-ezd-command
    `(scheme (repeat ,any?))
    "(scheme <scheme expression> ...)"
    (lambda args (for-each eval (car args))))

;;; The INCLUDE command loads ezd commands or Scheme expressions (files with
;;; a suffix of .sc) from a file.

(define (ezd-include file)
    (if (and (>= (string-length file) 3)
	     (equal? (substring file (- (string-length file) 3)
			 (string-length file))
		     ".sc"))
	(load file)
	(with-input-from-file
	    file
	    (lambda ()
		    (let loop ((exp (ezd-read)))
			 (unless (eof-object? exp)
				 (ezd exp)
				 (loop (ezd-read))))))))

(define-ezd-command
    `(include ,string?)
    "(include \"file-name\")"
    ezd-include)

;;; The QUIT command terminates ezd processing.

(define-ezd-command
    `(quit)
    "(quit)"
    (lambda ()
	    (let ((handling-events (and (isa-display? *display*)
					(display-handling-events *display*)))
		  (command-stream in-read-eval-draw))
		 (ezd-reset)
		 (if command-stream (set! ezd-done #t))
		 (if handling-events
		     (if (feature? 'csi)
			 (reset)
			 (exit))))))

;;; Command stepping.  Each step identifies itself by a STEP command with some
;;; expression.  The expression is assigned to *STEP* and then the value of
;;; *STEPPER* is evaluated.  If it is true, then the program pauses until some
;;; event calls NEXT-STEP or *STEPPER* is set to #f.

(define *step* #f)	;;; Value associated with this step.

(define *stepper* #f)	;;; Expression to evaluate at each step to decide
			;;; whether or not to continue.

(define *next-step* #f)	;;; Stepper advances when this is true.

;;; Define the stepper.

(define (stepper exp)
    (set! *stepper* exp))

(define-ezd-command
    `(stepper ,any?)
    "(stepper expression)"
    stepper)

;;; Test to see if the stepper requires a stop at this step.

(define (step exp)
    (set! *step* exp)
    (set! *next-step* #f)
    (when (eval *stepper*)
	  (display-event-handler *display*)
	  (let loop ()
	       (yselect *dpy* 1000000 0)
	       (display-event-handler *display*)
	       (if (and *stepper* (not *next-step*)) (loop)))))

(define-ezd-command
    `(step ,any?)
    "(step expression)"
    step)

;;; Call this from an event handler to allow further processing.

(define (next-step)
    (set! *next-step* #t))

;;; The PAUSE command waits for some number of milliseconds, or until ezd
;;; completes, before returning.

(define-ezd-command
    `(pause (optional ,positive-number?))
    "(pause [ millisecond pause time ])"
    (lambda (ms-pause)
	    (if ms-pause
		(thread-sleep! (/ ms-pause 1000))
		(wait-system-file #f))))

;;; Module reset/initialization procedure.

(define (ezd-module-init)
    (set! ezd-done #f))

)